lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -23,253 +23,308 @@
+  }
+
+public:
+  static Optional<ReferenceDependencyFileEmitter>
+  create(DiagnosticEngine &diags, SourceFile *SF,
+         const DependencyTracker &depTracker, StringRef outputPath) {
+    // Before writing to the dependencies file path, preserve any previous file
+    // that may have been there. No error handling -- this is just a nicety, it
+    // doesn't matter if it fails.
+    llvm::sys::fs::rename(outputPath, outputPath + "~");
+
+    std::error_code EC;
+    auto out = llvm::make_unique<llvm::raw_fd_ostream>(outputPath, EC,
+                                                       llvm::sys::fs::F_None);
+
+    if (out->has_error() || EC) {
+      diags.diagnose(SourceLoc(), diag::error_opening_output, outputPath,
+                     EC.message());
+      out->clear_error();
+      return None;
     }
+
+    return ReferenceDependencyFileEmitter(SF, depTracker, std::move(out));
   }
 
-  return VD->getFormalAccess() <= AccessLevel::FilePrivate;
-}
+  void emit() {
+    *out << "### Swift dependencies file v0 ###\n";
 
-static bool extendedTypeIsPrivate(TypeLoc inheritedType) {
-  auto type = inheritedType.getType();
-  if (!type)
-    return true;
+    emitProviders();
+    emitDependencies();
+    emitInterfaceHash();
+    }
 
-  if (!type->isExistentialType()) {
-    // Be conservative. We don't know how to deal with other extended types.
-    return false;
+    void emitProviders() const;
+    void emitDependencies() const;
+
+    void emitInterfaceHash() {
+      llvm::SmallString<32> interfaceHashVal;
+      SF->getInterfaceHash(interfaceHashVal);
+      *out << interfaceHash << ": \"" << interfaceHashVal << "\"\n";
   }
+};
 
-  auto layout = type->getExistentialLayout();
-  assert(!layout.superclass && "Should not have a subclass existential "
-         "in the inheritance clause of an extension");
-  for (auto protoTy : layout.getProtocols()) {
-    if (!declIsPrivate(protoTy->getDecl()))
-      return false;
+class ProviderEmitter {
+  llvm::MapVector<const NominalTypeDecl *, bool> extendedNominals;
+  llvm::SmallVector<const FuncDecl *, 8> memberOperatorDecls;
+  llvm::SmallVector<const ExtensionDecl *, 8> extensionsWithJustMembers;
+
+  llvm::raw_fd_ostream &out;
+
+public:
+  ProviderEmitter(llvm::raw_fd_ostream &out) : out(out) {}
+
+  void emit(const SourceFile *SF);
+
+private:
+  void emitTopLevelNames(const ArrayRef<const Decl *> decls);
+  void emitTopLevelDecl(const Decl *D);
+  void emitExtensionDecl(const ExtensionDecl *ED);
+  void emitNominalTypeDecl(const NominalTypeDecl *NTD);
+  void emitValueDecl(const ValueDecl *VD) const;
+
+  void emitNominalTypes() const;
+  void emitMembers() const;
+  void emitDynamicLookupMembers(const SourceFile *SF) const;
+
+  void findNominalsAndOperators(const DeclRange members);
+  static bool extendedTypeIsPrivate(const TypeLoc inheritedType);
+  static bool declIsPrivate(const Decl *member);
+
+  template <typename Decl_T> static bool hasVisibleName(const Decl_T *decl) {
```

Comment:
Why is this helper function a member function, but the other two top-level?

Reply:
`mangleTypeAsContext`, `hasSomePrivateAccessLevel`, and `isVisibleOutsideItsFile` are used by both `DependencyEmitter` and `ProviderEmitter`, so they cannot be members in either. Would a common superclass be better?

Reply:
Ah, I figured you could just make them all top-level, since that's what we usually do whenever the class declaration is in a header. But this is a reasonable explanation.